authz/providers/azure/rbac/rbac.go

```diff
@@ -0,0 +1,259 @@
+/*
+Copyright The Guard Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package rbac
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/appscode/guard/auth/providers/azure/graph"
+	"github.com/appscode/guard/authz"
+
+	"github.com/golang/glog"
+	"github.com/moul/http2curl"
+	"github.com/pkg/errors"
+	authzv1 "k8s.io/api/authorization/v1"
+)
+
+const (
+	managedClusters         = "Microsoft.ContainerService/managedClusters"
+	connectedClusters       = "Microsoft.Kubernetes/connectedClusters"
+	checkAccessPath         = "/providers/Microsoft.Authorization/checkaccess"
+	checkAccessAPIVersion   = "2018-09-01-preview"
+	remaingSubReadARMHeader = "x-ms-ratelimit-remaining-subscription-reads"
+	expiryDelta             = 60 * time.Second
+)
+
+type void struct{}
+
+// AccessInfo allows you to check user access from MS RBAC
+type AccessInfo struct {
+	headers   http.Header
+	client    *http.Client
+	expiresAt time.Time
+	// These allow us to mock out the URL for testing
+	apiURL *url.URL
+
+	tokenProvider            graph.TokenProvider
+	clusterType              string
+	azureResourceId          string
+	armCallLimit             int
+	dataStore                authz.Store
+	skipCheck                map[string]void
+	retrieveGroupMemberships bool
+	skipAuthzForNonAADUsers  bool
+}
+
+func newAccessInfo(tokenProvider graph.TokenProvider, rbacURL *url.URL, clsuterType, resourceId string, armCallLimit int, dataStore authz.Store, skipList []string, retrieveGroupMemberships, skipAuthzForNonAADUsers bool) (*AccessInfo, error) {
+	u := &AccessInfo{
```

Comment:
I think functional options will be a good idea here. https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis